### ISSUE TITLE
Could not decorrelate subquery inside of `CASE` statement

### ISSUE THREAD

**Issue by benjie** [cockroachdb/cockroach#67951]
**Describe the problem**

```
ERROR:  XXUUU: could not decorrelate subquery
LOCATION:  decorrelationError, builder.go:196
```

**To Reproduce**

1. Run cockroach with this script: https://github.com/graphile/graphile-engine/blob/7e7171a2db14c8011e39e30b325df6b9ea08ed3c/packages/pg-analogues/cockroach/run

2. Set up the following database schema:

```sql
create schema c;
create table c.person (
  id serial primary key,
  person_full_name varchar not null,
  aliases text[] not null default '{}',
  about text,
  email /* b.email */ text not null unique,
  site /* b.wrapped_url */ text default null,
  config /* hstore */ json,
  last_login_from_ip inet,
  last_login_from_subnet /* cidr */ text,
  user_mac /* macaddr */ text,
  created_at timestamp default current_timestamp
);
insert into c.person (id, person_full_name, email, about, config, last_login_from_ip, last_login_from_subnet, user_mac, created_at) values
  (1, 'John Smith', 'john.smith@email.com', null, '{}', '192.168.0.1', '192.168.0.0/24', '00:00:00:00:00:00', null),
  (2, 'Sara Smith', 'sara.smith@email.com', null, '{"a":1}', '10.1.2.3', '10.0.0.0/8', '11-22-33-44-55-66', null),
  (3, 'Budd Deey', 'budd.deey@email.com', 'Just a friendly human', '{"b":2}', '172.21.22.23', '172.16.0.0/12', '778899AABBCC', null),
  (4, 'Kathryn Ramirez', 'kathryn.ramirez@email.com', null, '{"a": 3, "b": 4}', '192.168.1.2', '192.168.0.0/16', 'ddee.ff00.1122', null),
  (5, 'Joe Tucker', 'joe.tucker@email.com', null, '{"a": 5, "b": 6, "actually_null": null, "null_string": "null", "_\"_$@£$)(@*£$": "_\"_$@£$)(@*£$"}', '192.168.0.2', '192.168.0.0/23', '334455-667788', null),
  (6, 'Twenty Seventwo', 'graphile-build.issue.27.exists@example.com', null, '{"null_1": null, "null_2": null}', '10.1.2.254', '10.0.0.0/9', '99aabb:ccddee', null);
```

3. Issue the following query

```sql
select case
  when array_length(person."aliases", 1) = 0
    then '[]'::json
  else (
    select json_agg(tmp3)
    from unnest(person."aliases") as tmp3
  )
  end
from c.person;
```

<details>
<summary>Note the above is simplified, original query is inside this (click me).</summary>

Here's the query PostGraphile is issuing that Cockroach cannot decorrelate:

```sql
with __local_0__ as (
  select to_json(
    (
      json_build_object(
        'aliases'::text,
        (
          case when (__local_1__."aliases") is null then null when coalesce(
            array_length(
              (__local_1__."aliases"),
              1
            ),
            0
          ) = 0 then '[]'::json else (
            select json_agg(__local_2__)
            from unnest((__local_1__."aliases")) as __local_2__
          ) end
        )
      )
    )
  ) as "@nodes"
  from (
    select __local_1__.*
    from "c"."person" as __local_1__
    where (TRUE) and (TRUE)
    order by __local_1__."id" ASC
    limit 1
  ) __local_1__
),
__local_3__ as (
  select json_agg(
    to_json(__local_0__)
  ) as data
  from __local_0__
)
select coalesce(
  (
    select __local_3__.data
    from __local_3__
  ),
  '[]'::json
) as "data"
```

</details>

```
ERROR:  XXUUU: could not decorrelate subquery
LOCATION:  decorrelationError, builder.go:196
```

**Expected behavior**

Postgres outputs nulls:

```
select case
  when array_length(person."aliases", 1) = 0
    then '[]'::json
  else (
    select json_agg(tmp3)
    from unnest(person."aliases") as tmp3
  )
  end
from c.person;
┌──────────┐
│ json_agg │
├──────────┤
│ ¤        │
│ ¤        │
│ ¤        │
│ ¤        │
│ ¤        │
│ ¤        │
└──────────┘
(6 rows)

Time: 0.735 ms
```

**Environment:**
 - CockroachDB version:  CCL v21.2.0-alpha.00000000-2606-g5c56090559 @ 2021/07/22 16:59:03 (go1.16.6)
 - Server OS: Ubuntu 21.04
 - Client app: psql, node-postgres

**Additional context**
What was the impact: PostGraphile cannot run against (some) arrays.

Epic CRDB-2410

Jira issue: CRDB-8798

**Comment by RaduBerinde** [comment]
I think there are some limitations with decorrelating queries around  `CASE`. The crux of the matter is that the semantics of `CASE` are such that the branches not taken are not evaluated (and thus can't error out).

The variant below seems to work. I believe this is equivalent, what do you think?

```
          select coalesce(json_agg(__local_2__), if((__local_1__."aliases") is null, null, '[]'::jsonb))
          from unnest((__local_1__."aliases")) as __local_2__
```

```
with __local_0__ as (
  select to_json(
    (
      json_build_object(
        'aliases'::text,
        (
          select coalesce(json_agg(__local_2__), if((__local_1__."aliases") is null, null, '[]'::jsonb))
          from unnest((__local_1__."aliases")) as __local_2__
        )
      )
    )
  ) as "@nodes"
  from (
    select __local_1__.*
    from "c"."person" as __local_1__
    where (TRUE) and (TRUE)
    order by __local_1__."id" ASC
    limit 1
  ) __local_1__
),
__local_3__ as (
  select json_agg(
    to_json(__local_0__)
  ) as data
  from __local_0__
)
select coalesce(
  (
    select __local_3__.data
    from __local_3__
  ),
  '[]'::json
) as "data"
```

**Comment by RaduBerinde** [comment]
For the smaller query that was used as an example, this is the variant that works:
```sql
SELECT
  (
    SELECT
      IF(array_length(person.aliases, 1) = 0, '[]'::JSONB, json_agg(tmp3))
    FROM
      unnest(person.aliases) AS tmp3
  )
FROM
  c.person
```

**Comment by benjie** [comment]
The following indeed gives the correct results:

```sql
create table person (id serial primary key, aliases text[]); 

insert into person (aliases) values                                                                                      
  (null),
  (array[]::text[]),
  (array['bob', 'robbie']::text[]);

select
  json_build_object(
    'id', id,
    'l', array_length(aliases, 1),
    'aliases', (
      select
        case
          when aliases is null then null
          when coalesce(array_length(aliases, 1), 0) = 0 then '[]'::json
          else json_agg(a)
        end as json
      from unnest(aliases) a
    )
  )
from c.person;
```

```
┌────────────────────────────────────────────────────┐
│                     cockroach                      │
├────────────────────────────────────────────────────┤
│ {"id" : 1, "l" : null, "aliases" : null}           │
│ {"id" : 2, "l" : null, "aliases" : []}             │
│ {"id" : 3, "l" : 2, "aliases" : ["bob", "robbie"]} │
└────────────────────────────────────────────────────┘
```

Turns out that `array_length(x, 1)` returns null when `x` has length 0 (unlike `cardinality` which cockroach does not yet support) so had to add an extra coalesce to get the right results. And we can't use `if` since Postgres doesn't have that function.

**Comment by benjie** [comment]
Turns out PostgreSQL and Cockroach handle empty arrays differently in this case; given the following query:

```sql
select (
  select
    case
      when (__local_1__."aliases") is null
      then null
      when coalesce(
        array_length(__local_1__."aliases", 1),
        0
      ) = 0
      then '[]'::json
      else json_agg(__local_2__)
    end
  from unnest(__local_1__."aliases") __local_2__
)
from (
  select array[]::text[] as "aliases"
) as __local_1__;
```

Cockroach returns `null` whereas Postgres returns `"[]"::json`. I believe the original `case` statement where the `unnest` only occurred inside the `else` clause is more correct.

**Comment by benjie** [comment]
I've updated the title, I think it's the `case` statement that's the issue here. Here's another example that's been surfaced from PostGraphile's tests:

```sql
select
  (
    select
      case
        when (__local_1__ is null)
        then null
        else json_build_object(
          'personId'::text, __local_1__."person_id",
          '@personByPersonId'::text, (
            select json_build_object(
              'id'::text,
              __local_2__."id"
            ) as object
            from "c"."person" as __local_2__
            where __local_1__."person_id" = __local_2__."id"
          )
        )
      end as object
    from "c"."person_secret" as __local_1__
    where __local_1__."person_id" = __local_0__."id"
  ) as "@secret"
from "c"."person" as __local_0__
where __local_0__."id" = 2;
```

PostgreSQL outputs:

```
┌────────────────────────────────────────────────────┐
│                      @secret                       │
├────────────────────────────────────────────────────┤
│ {"personId" : 2, "@personByPersonId" : {"id" : 2}} │
└────────────────────────────────────────────────────┘
(1 row)
```

Cockroach:

```
ERROR:  XXUUU: could not decorrelate subquery
LOCATION:  decorrelationError, builder.go:196
```

Seed files (for structure and data):

- https://github.com/graphile/graphile-engine/blob/74eb00f4614dffceb4c2c4443207a18919796dd3/scripts/pretest#L64-L92
- https://github.com/graphile/graphile-engine/blob/74eb00f4614dffceb4c2c4443207a18919796dd3/packages/pg-analogues/__tests__/kitchen-sink-schema.analogue.sql
- https://github.com/graphile/graphile-engine/blob/74eb00f4614dffceb4c2c4443207a18919796dd3/packages/pg-analogues/__tests__/kitchen-sink-schema-inserts.analogue.sql
- https://github.com/graphile/graphile-engine/blob/74eb00f4614dffceb4c2c4443207a18919796dd3/packages/pg-analogues/__tests__/kitchen-sink-data.analogue.sql

**Comment by RaduBerinde** [comment]
Yes, it's the shortcutting aspect of CASE that prevents decorrelation in these cases. Switching to `if` I think would fix it but that is not available in `postgres`.

> Turns out PostgreSQL and Cockroach handle empty arrays differently in this case; given the following query:
> 
> ```sql
> select (
>   select
>     case
>       when (__local_1__."aliases") is null
>       then null
>       when coalesce(
>         array_length(__local_1__."aliases", 1),
>         0
>       ) = 0
>       then '[]'::json
>       else json_agg(__local_2__)
>     end
>   from unnest(__local_1__."aliases") __local_2__
> )
> from (
>   select array[]::text[] as "aliases"
> ) as __local_1__;
> ```
> 
> Cockroach returns `null` whereas Postgres returns `"[]"::json`. I believe the original `case` statement where the `unnest` only occurred inside the `else` clause is more correct.

I'm kind of confused by this query. If `aliases` is an empty array, wouldn't we expect no rows from the `SELECT ... FROM unnest(aliases)` (in both pg and crdb)?  Maybe you can come up with a simplified example that can be run easily?

**Comment by benjie** [comment]
I think the json_agg being present in the selection automatically turns it into an aggregate query even though it’s not actually used, so there’s one row from the aggregate. Cockroach does the same: returns one row (rather than no rows), just the row contains null.

Please note this query is effectively what you suggested [above](https://github.com/cockroachdb/cockroach/issues/67951#issuecomment-885200835) but with CASE instead of IF.

**Comment by benjie** [comment]
> > ```sql
> > select (
> >   select
> >     case
> >       when (__local_1__."aliases") is null
> >       then null
> >       when coalesce(
> >         array_length(__local_1__."aliases", 1),
> >         0
> >       ) = 0
> >       then '[]'::json
> >       else json_agg(__local_2__)
> >     end
> >   from unnest(__local_1__."aliases") __local_2__
> > )
> > from (
> >   select array[]::text[] as "aliases"
> > ) as __local_1__;
> > ```
> Maybe you can come up with a simplified example that can be run easily?

That _is_ the simplified example and it can be ran standalone, doesn't even require any data or tables. Just copy/paste it into any Cockroach (or PostgreSQL) `psql` instance and you should see the behaviour.

**Comment by benjie** [comment]
Regarding this query:

```sql
select
  (
    select
      case
        when (__local_1__ is null)
        then null
        else json_build_object(
          'personId'::text, __local_1__."person_id",
          '@personByPersonId'::text, (
            select json_build_object(
              'id'::text,
              __local_2__."id"
            ) as object
            from "c"."person" as __local_2__
            where __local_1__."person_id" = __local_2__."id"
          )
        )
      end as object
    from "c"."person_secret" as __local_1__
    where __local_1__."person_id" = __local_0__."id"
  ) as "@secret"
from "c"."person" as __local_0__
where __local_0__."id" = 2;
```

interestingly you **cannot** reproduce it if you replace the tables with `values` subqueries:

```sql
select
  (
    select
      case
        when (__local_1__ is null)
        then null
        else json_build_object(
          'personId'::text, __local_1__."person_id",
          '@personByPersonId'::text, (
            select json_build_object(
              'id'::text,
              __local_2__."id"
            ) as object
            from (values (2)) as __local_2__ (id)
            where __local_1__."person_id" = __local_2__."id"
          )
        )
      end as object
    from (values (2)) as __local_1__ (person_id)
    where __local_1__."person_id" = __local_0__."id"
  ) as "@secret"
from (values (2)) as __local_0__ (id)
where __local_0__."id" = 2;
```

In this case it works fine! :astonished: 

Here's a full reproduction that can be ran against an empty database:

```sql
create schema c;
create table c.person (id int primary key);
create table c.person_secret(person_id int primary key);
insert into c.person(id) values (2);
insert into c.person_secret (person_id) values (2);

select
  (
    select
      case
        when (__local_1__ is null)
        then null
        else json_build_object(
          'personId'::text, __local_1__."person_id",
          '@personByPersonId'::text, (
            select json_build_object(
              'id'::text,
              __local_2__."id"
            ) as object
            from "c"."person" as __local_2__
            where __local_1__."person_id" = __local_2__."id"
          )
        )
      end as object
    from "c"."person_secret" as __local_1__
    where __local_1__."person_id" = __local_0__."id"
  ) as "@secret"
from "c"."person" as __local_0__
where __local_0__."id" = 2;
```

Produces:

```
ERROR:  XXUUU: could not decorrelate subquery
LOCATION:  decorrelationError, builder.go:196
```

Note that this one does not involve aggregates.

**Comment by RaduBerinde** [comment]
> That is the simplified example and it can be ran standalone, doesn't even require any data or tables. Just copy/paste it into any Cockroach (or PostgreSQL) psql instance and you should see the behaviour.

> I think the json_agg being present in the selection automatically turns it into an aggregate query even though it’s not actually used, so there’s one row from the aggregate. Cockroach does the same: returns one row (rather than no rows), just the row contains null.

Ah, sorry, thanks! The construct of `CASE` which involves a simple expression in some branches and an aggregate in another branch blows my mind, but I see it now. It does look like a bug.

**Comment by RaduBerinde** [comment]
> Ah, sorry, thanks! The construct of CASE which involves a simple expression in some branches and an aggregate in another branch blows my mind, but I see it now. It does look like a bug.

Filed https://github.com/cockroachdb/cockroach/issues/68290 for that issue.

**Comment by mgartner** [comment]
This was fixed by #95234, which should be included in v23.1. Closing.
